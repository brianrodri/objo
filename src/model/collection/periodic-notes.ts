import { notStrictEqual as assertNotStrictEqual } from "assert";
import { attempt, clone, isError } from "lodash";
import { DateTime, DateTimeOptions, Duration, DurationLike, Interval, IntervalMaybeValid } from "luxon";
import { parse } from "path";
import { DeepReadonly } from "utility-types";

import { assertLuxonFormat, assertValid } from "@/util/luxon-utils";

import { DateBasedCollection } from "./schema";
import { stripTrailingSlash } from "./util";

/**
 * Intended to handle the popular "Periodic Notes" community plugin.
 * As a consequence, all files in the collection must be placed in the same folder.
 * @see {@link https://github.com/liamcain/obsidian-periodic-notes}
 */
export class PeriodicNotes extends DateBasedCollection implements PeriodicNotesConfig<true> {
    /** The folder containing all of the notes. TODO: Support files organized into different folders. */
    public readonly folder: string;

    /**
     * The format used to parse a {@link DateTime} from file names. Must work for both parsing and formatting.
     * @see {@link https://moment.github.io/luxon/#/parsing?id=table-of-tokens}
     */
    public readonly dateFormat: string;

    /** Luxon options used when parsing {@link DateTime}s from file names. */
    public readonly dateOptions: DeepReadonly<DateTimeOptions>;

    /**
     * The {@link Duration} of each file's corresponding {@link Interval}.
     * @example
     * Daily notes should use a duration of `{ days: 1 }`.
     */
    public readonly intervalDuration: Duration<true>;

    /**
     * Offset between the file's _parsed_ date and the corresponding {@link Interval}'s _start_ date. May be negative.
     * @example
     * Sprint notes may use ISO weeks as their {@link dateFormat}, for example: `2025-W10.md`.
     * If sprints _actually_ begin on Thursdays rather than Mondays, then we can adjust the interval with `{ days: 3 }`.
     */
    public readonly intervalOffset: Duration<true>;

    public constructor(config: PeriodicNotesConfig<boolean>) {
        super();
        const { folder, dateFormat, dateOptions, intervalDuration, intervalOffset } = validated(config);
        this.folder = folder;
        this.dateFormat = dateFormat;
        this.dateOptions = dateOptions;
        this.intervalDuration = intervalDuration;
        this.intervalOffset = intervalOffset;
    }

    /**
     * Used to accurately enable date-based queries on collections and to look up "neighboring" files.
     * @param filePath - the path to check.
     * @returns the {@link Interval} corresponding to the file, otherwise an invalid interval.
     */
    public override getIntervalOf(filePath: string): IntervalMaybeValid {
        const parsedPath = parse(filePath);
        if (parsedPath.dir !== this.folder) {
            return Interval.invalid("invalid folder", `"${filePath}" is not in "${this.folder}"`);
        }
        const parsedDate = DateTime.fromFormat(parsedPath.name, this.dateFormat, this.dateOptions);
        if (!parsedDate.isValid) {
            return Interval.invalid("invalid filename", `${parsedDate.invalidReason} ${parsedDate.invalidExplanation}`);
        }
        return Interval.after(parsedDate.plus(this.intervalOffset), this.intervalDuration);
    }
}

/**
 * Configuration options for {@link PeriodicNotes}.
 * @typeParam IsValidConfiguration - whether the configuration is valid.
 */
export type PeriodicNotesConfig<IsValidConfiguration extends boolean> =
    IsValidConfiguration extends true ?
        {
            /** {@inheritDoc PeriodicNotes.folder} */
            folder: string;
            /** {@inheritDoc PeriodicNotes.dateFormat} */
            dateFormat: string;
            /** {@inheritDoc PeriodicNotes.dateOptions} */
            dateOptions: DateTimeOptions;
            /** {@inheritDoc PeriodicNotes.intervalDuration} */
            intervalDuration: Duration<true>;
            /** {@inheritDoc PeriodicNotes.intervalOffset} */
            intervalOffset: Duration<true>;
        }
    :   {
            /** {@inheritDoc PeriodicNotes.folder} */
            folder: string;
            /** {@inheritDoc PeriodicNotes.dateFormat} */
            dateFormat: string;
            /** {@inheritDoc PeriodicNotes.dateOptions} */
            dateOptions?: DateTimeOptions;
            /** {@inheritDoc PeriodicNotes.intervalDuration} */
            intervalDuration: DurationLike;
            /** {@inheritDoc PeriodicNotes.intervalOffset} */
            intervalOffset?: DurationLike;
        };

/**
 * Validates and normalizes a PeriodicNotes configuration object.
 *
 * This function checks that:
 * - the folder path (after stripping any trailing slash) is non-empty,
 * - the date format is valid with the given date options,
 * - the interval duration is a valid, non-zero duration, and
 * - the interval offset is a valid duration.
 *
 * If any of these validations fail, an {@link AggregateError} is thrown containing details
 * about each issue. On success, a new configuration object with strictly defined properties is returned.
 * @param config - The configuration object to validate.
 * @returns A validated configuration object with normalized properties.
 * @throws If one or more configuration properties are invalid.
 */
function validated(config: PeriodicNotesConfig<false>): PeriodicNotesConfig<true> {
    const folder = stripTrailingSlash(config.folder);
    const dateFormat = config.dateFormat;
    const dateOptions = clone(config.dateOptions ?? {});
    const intervalDuration = Duration.fromDurationLike(config.intervalDuration);
    const intervalOffset = Duration.fromDurationLike(config.intervalOffset ?? 0);

    const errors = [
        attempt(() => assertNotStrictEqual(folder.length, 0, "folder must be non-empty")),
        attempt(() => assertLuxonFormat(dateFormat, dateOptions)),
        attempt(() => assertValid(intervalDuration, "interval duration is invalid")),
        attempt(() => assertNotStrictEqual(intervalDuration.valueOf(), 0, "interval duration must not be zero")),
        attempt(() => assertValid(intervalOffset, "interval offset is invalid")),
    ].filter(isError);

    if (errors.length > 0) {
        const indentedMessages = errors.map((error) => `${error.name}: ${error.message}`.replace(/^/gm, "\t"));
        throw new AggregateError(errors, ["invalid config", ...indentedMessages].join("\n"));
    }

    return { folder, dateFormat, dateOptions, intervalDuration, intervalOffset };
}
