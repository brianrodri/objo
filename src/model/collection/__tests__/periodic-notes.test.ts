import { Duration, DurationLike, Interval } from "luxon";
import { describe, expect, it } from "vitest";

import { PeriodicNotes, PeriodicNotesConfig } from "../periodic-notes";

describe(`${PeriodicNotes.name}`, () => {
    const required: PeriodicNotesConfig<false> = {
        folder: "/vault",
        dateFormat: "yyyy-MM-dd",
        intervalDuration: { days: 1 },
    };

    describe("pre-conditions", () => {
        it("should throw when folder is empty", () => {
            const folder = "";
            expect(() => new PeriodicNotes({ ...required, folder })).toThrowErrorMatchingSnapshot();
        });

        it("should throw when date format is empty", () => {
            const dateFormat = "";
            expect(() => new PeriodicNotes({ ...required, dateFormat })).toThrowErrorMatchingSnapshot();
        });

        it("should throw when interval duration is invalid", () => {
            const intervalDuration = { days: "one" } as unknown as DurationLike;
            expect(() => new PeriodicNotes({ ...required, intervalDuration })).toThrowErrorMatchingSnapshot();
        });

        it("should throw when interval offset is invalid", () => {
            const intervalOffset = { days: "one" } as unknown as DurationLike;
            expect(() => new PeriodicNotes({ ...required, intervalOffset })).toThrowErrorMatchingSnapshot();
        });

        it("should throw when interval duration is 0", () => {
            const intervalDuration = 0;
            expect(() => new PeriodicNotes({ ...required, intervalDuration })).toThrowErrorMatchingSnapshot();
        });

        it("should throw when everything is invalid", () => {
            const invalidConfig = {
                folder: "",
                dateFormat: "",
                intervalDuration: Duration.invalid("!"),
                intervalOffset: Duration.invalid("invalid time", "unspecified"),
            };
            expect(() => new PeriodicNotes(invalidConfig)).toThrowErrorMatchingSnapshot();
        });

        it.each([+1, 0, -1])("should not throw when interval offset is %j", (intervalOffset) => {
            expect(() => new PeriodicNotes({ ...required, intervalOffset })).not.toThrow();
        });

        it.each([+1, -1])("should not throw when interval duration is %j", (intervalDuration) => {
            expect(() => new PeriodicNotes({ ...required, intervalDuration })).not.toThrow();
        });
    });

    describe("post-conditions", () => {
        it("should strip trailing slashes from the folder", () => {
            expect(new PeriodicNotes({ ...required, folder: "/vault" }).folder).toEqual("/vault");
        });
    });

    describe('given daily log in folder: "/vault"', () => {
        const log = new PeriodicNotes({
            folder: "/vault",
            dateFormat: "yyyy-MM-dd",
            intervalDuration: { days: 1 },
        });

        it.each(["/vault/2023-01-01.md"])("should include %j", (filePath) => {
            expect(log.includes(filePath)).toBe(true);
        });

        it.each([
            "/vault/2023-01-99.md",
            "/vault/2023/01/99.md",
            "/vault/subfolder/2023-01-04.md",
            "/sprints/2023-01-05.md",
        ])("should not include %j", (filePath) => {
            expect(log.includes(filePath)).toBe(false);
        });

        it.each([
            ["2023-01-01/2023-01-02", "/vault/2023-01-01.md"],
            ["2025-02-28/2025-03-01", "/vault/2025-02-28.md"],
        ])("should return %j as the interval of %j", (isoInterval, filePath) => {
            expect(log.getIntervalOf(filePath)).toEqual(Interval.fromISO(isoInterval));
        });
    });

    describe('given sprint logs starting every other thursday in folder: "/sprints"', () => {
        const log = new PeriodicNotes({
            folder: "/sprints",
            dateFormat: "kkkk-'W'WW",
            intervalDuration: { weeks: 2 },
            intervalOffset: { days: 3 },
        });

        it.each(["/sprints/2023-W37.md"])("should include %j", (filePath) => {
            expect(log.includes(filePath)).toBe(true);
        });

        it.each([
            "/sprints/2023-09-14.md",
            "/sprints/2023-W99.md",
            "/sprints/subfolder/2023-W37.md",
            "/vault/2023-W37.md",
        ])("should not include %j", (filePath) => {
            expect(log.includes(filePath)).toBe(false);
        });

        it.each([
            ["2023-09-14/2023-09-28", "/sprints/2023-W37.md"],
            ["2025-02-27/2025-03-13", "/sprints/2025-W09.md"],
        ])("should return %j as the interval of %j", (isoInterval, filePath) => {
            expect(log.getIntervalOf(filePath)).toEqual(Interval.fromISO(isoInterval));
        });
    });
});
