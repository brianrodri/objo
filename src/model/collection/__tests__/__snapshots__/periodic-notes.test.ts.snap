// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PeriodicNotes > pre-conditions > should throw when date format is empty 1`] = `
[AggregateError: invalid config
	AssertionError: date format must be non-empty]
`;

exports[`PeriodicNotes > pre-conditions > should throw when everything is invalid 1`] = `
[AggregateError: invalid config
	AssertionError: folder must be non-empty
	AssertionError: date format must be non-empty
	AssertionError: interval offset is invalid: invalid time: unspecified
	AssertionError: interval duration is invalid: invalid time: infinity]
`;

exports[`PeriodicNotes > pre-conditions > should throw when folder is empty 1`] = `
[AggregateError: invalid config
	AssertionError: folder must be non-empty]
`;

exports[`PeriodicNotes > pre-conditions > should throw when interval duration is 0 1`] = `
[AggregateError: invalid config
	AssertionError: interval duration must not be zero]
`;

exports[`PeriodicNotes > pre-conditions > should throw when interval duration is invalid 1`] = `[Error: Invalid unit value one]`;

exports[`PeriodicNotes > pre-conditions > should throw when interval offset is invalid 1`] = `[Error: Invalid unit value one]`;
